File: go/Assembler/engine/preprocessjson/engine_preprocess_json.go
package engine_preprocessjson

import (
	Logger "arshu/common"
	interfaces "assembler/interface"
	"assembler/model"
	"fmt"
)

type EnginePreProcessJson struct {
	AppViewPrefix string
}

func NewEnginePreProcessJson(prefix string) *EnginePreProcessJson {
	return &EnginePreProcessJson{AppViewPrefix: prefix}
}

func (e *EnginePreProcessJson) MergeTemplates(appSite, appFile, appView string, loader interfaces.ILoaderJson, enableJsonProcessing bool) string {
	Logger.Debug(fmt.Sprintf("MergeTemplates called: appSite=%s, appFile=%s, appView=%s, enableJson=%t", appSite, appFile, appView, enableJsonProcessing), "EnginePreProcessJson")

	if loader == nil {
		Logger.Warn("No loader provided", "EnginePreProcessJson")
		return ""
	}

	// Use ILoaderJson to retrieve the main template
	result := loader.GetTemplateHtml(appSite, appFile, appView, e.AppViewPrefix)
	if result == nil {
		Logger.Warn(fmt.Sprintf("Main template not found for appSite=%s, appFile=%s", appSite, appFile), "EnginePreProcessJson")
		return ""
	}

	mainPreprocessed, ok := result.(*model.PreprocessedTemplate)
	if !ok || mainPreprocessed == nil {
		Logger.Warn(fmt.Sprintf("Main template not found or invalid type for appSite=%s, appFile=%s", appSite, appFile), "EnginePreProcessJson")
		return ""
	}

	Logger.Debug(fmt.Sprintf("Main template found, original size: %d", len(mainPreprocessed.OriginalContent)), "EnginePreProcessJson")

	// Start with original content
	contentHtml := mainPreprocessed.OriginalContent

	// Merge JSON into main template first using loader's centralized method
	if enableJsonProcessing {
		contentHtml = loader.MergeHtmlWithJson(contentHtml, appSite, appFile)
		Logger.Debug(fmt.Sprintf("After main template JSON merge: %d chars", len(contentHtml)), "EnginePreProcessJson")
	}

	// Apply ALL replacement mappings from ALL templates using loader's method
	contentHtml = loader.ApplyAllReplacementMappings(contentHtml, appSite, mainPreprocessed, appView, e.AppViewPrefix, enableJsonProcessing)

	Logger.Debug(fmt.Sprintf("MergeTemplates complete: output size=%d", len(contentHtml)), "EnginePreProcessJson")
	return contentHtml
}

File: php/Assembler/src/Engine/PreProcessJson/EnginePreProcessJson.php
<?php

namespace Assembler\Engine\PreProcessJson;

use Assembler\Interface\ILoaderJson;

use Arshu\Common\Logger;
use Assembler\Loader\LoaderPreProcessJson;

class EnginePreProcessJson
{
    private string $appViewPrefix = '';

    public function setAppViewPrefix(string $prefix): void
    {
        $this->appViewPrefix = $prefix;
    }

    public function mergeTemplates(string $appSite, string $appFile, ?string $appView, ILoaderJson $loader, bool $enableJsonProcessing = true): string
    {
        Logger::debug("MergeTemplates called: appSite=$appSite, appFile=$appFile, appView=" . ($appView ?? 'null') . ", enableJson=$enableJsonProcessing", 'EnginePreProcessJson');

        if (!$loader) {
            Logger::warn('No loader provided', 'EnginePreProcessJson');
            return '';
        }

        // Use ILoaderJson to retrieve the main template
        $mainPreprocessed = $loader->getTemplateHtml($appSite, $appFile, $appView, $this->appViewPrefix);
        if (!$mainPreprocessed) {
            Logger::warn("Main template not found for appSite=$appSite, appFile=$appFile", 'EnginePreProcessJson');
            return '';
        }

        Logger::debug('Main template found, original size: ' . strlen($mainPreprocessed->getOriginalContent()), 'EnginePreProcessJson');

        // Start with original content
        $contentHtml = $mainPreprocessed->getOriginalContent();

        // Merge JSON into main template first using loader's centralized method
        if ($enableJsonProcessing) {
            $contentHtml = $loader->mergeHtmlWithJson($contentHtml, $appSite, $appFile);
            Logger::debug('After main template JSON merge: ' . strlen($contentHtml) . ' chars', 'EnginePreProcessJson');
        }

        // Apply ALL replacement mappings from ALL templates using loader's method
        $contentHtml = $loader->applyAllReplacementMappings($contentHtml, $appSite, $mainPreprocessed, $appView, $this->appViewPrefix, $enableJsonProcessing);

        Logger::debug('MergeTemplates complete: output size=' . strlen($contentHtml), 'EnginePreProcessJson');
        return $contentHtml;
    }
}